### PR TITLE
[Experiment] what if reorder compare of columns in Merge

### PR DESCRIPTION
**Pull request by UnamedRus**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Start to compare rows from most likely unequal columns first.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit bb5fde6220449c3776524ceac568e651eaaa1d3f with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63780/bb5fde6220449c3776524ceac568e651eaaa1d3f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
Did it help?

**Comment by UnamedRus** [comment]
> Did it help?

Doesn't seems so.
But may be, i was bad on making good test for that.

**Comment by alexey-milovidov** [comment]
We need at least one proof, even synthetic will be ok.

**Comment by UnamedRus** [comment]
Test with LC(String)
```
New 
SELECT count()
FROM test_reorder_columns_in_merge
FINAL
SETTINGS max_threads = 1

   ┌──count()─┐
1. │ 60000000 │ -- 60.00 million
   └──────────┘

1 row in set. Elapsed: 3.120 sec. Processed 59.91 million rows, 479.32 MB (19.20 million rows/s., 153.63 MB/s.)
Peak memory usage: 4.06 MiB.

Old
SELECT count()
FROM test_reorder_columns_in_merge
FINAL
SETTINGS max_threads = 1

   ┌──count()─┐
1. │ 60000000 │ -- 60.00 million
   └──────────┘

1 row in set. Elapsed: 4.134 sec. Processed 59.91 million rows, 479.32 MB (14.49 million rows/s., 115.94 MB/s.)
Peak memory usage: 4.06 MiB.
```

File: src/Processors/Merges/Algorithms/RowRef.h [hunk 1]
```diff
@@ -159,6 +159,26 @@ struct RowRef
         return true;
     }
 
+    static size_t checkEqualsFirstNonEqual(size_t size, size_t offset, const IColumn ** lhs, size_t lhs_row, const IColumn ** rhs, size_t rhs_row)
+    {
+        for (size_t i = 0; i < size; ++i)
+        {
+            auto col_number = (offset + i) % size;
```
Comment (alexey-milovidov) [inline src/Processors/Merges/Algorithms/RowRef.h:166]: I'm afraid of this division - change to something like if (offset + i == size) offset = 0

Comment (UnamedRus) [reply src/Processors/Merges/Algorithms/RowRef.h:166]: Not sure that simple reset to 0 will work

But something like, could.
But not sure that it's better

```
if (offset + i) == size)
   offset -= size

auto col_number = (offset + i)
```

**Pull request merged**